oozie-to-airflow/tests/test_templates.py

```diff
@@ -0,0 +1,391 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests Templates"""
+import ast
+import unittest
+from copy import deepcopy
+from random import randint
+from typing import Dict, Any
+
+from parameterized import parameterized
+from airflow.utils.trigger_rule import TriggerRule
+
+from converter.primitives import Task, Relation
+from utils.template_utils import render_template
+
+DELETE_MARKER: Any = {}
+
+
+def mutate(parent: Dict[str, Any], mutations: Dict[str, Any]):
+    result = {}
+    result.update(parent)
+    for key, value in mutations.items():
+        if value is DELETE_MARKER:
+            del result[key]
+        elif isinstance(value, dict):
+            result[key] = mutate(parent.get(key, {}), value)
+        else:
+            result[key] = value
+    return result
+
+
+def get_by_path(target, segments):
+    result = target
+    for segment in segments:
+        if isinstance(result, dict):
+            result = result[segment]
+        elif isinstance(result, list):
+            result = result[int(segment)]
+        else:
+            raise Exception(f"Invalid path: {'.'.join(segments)}")
+    return result
+
+
+def set_by_path(target, segments, value):
+    result = get_by_path(target, segments[:-1])
+    if isinstance(result, dict):
+        result[segments[-1]] = value
+    elif isinstance(result, list):
+        result[int(segments[-1])] = value
+    else:
+        raise Exception(f"Invalid path: {'.'.join(segments)}")
+
+
+class TemplateTestMixin:
+    # pylint: disable=invalid-name
+    def assertValidPython(self, code):
+        self.assertTrue(ast.parse(code))
+
+    def test_all_template_parameters_must_be_correlated_with_output(self):
+        original_view = render_template(self.TEMPLATE_NAME, **self.DEFAULT_TEMPLATE_PARAMS)
+
+        def mutate_random(previous_segments):
```

Comment:
This function name is not very informative. Could we change it to make it clearer what happens here?

Reply:
I added a description of this test.

Reply:
It's a good description, thanks. That said I think the method could still be a bit more verbose.

Reply:
Do you have suggestions for this name? I had no idea. I decided that if I add a description to the test, it is not necessary to change the name. This function can only be called from this block of code. This code block consists of definitions and function calls.  It seems to me that everything in this piece of code is already understandable.

Reply:
I will think about it, but you may want to take a look at Uncle Bob's short article that I stumbled upon recently which explains why such a function should have a descriptive name http://www.informit.com/articles/article.aspx?p=1323426

Reply:
I think everything is okay. I invite you to send another PR, which will change the name of this method.

Reply:
My article does not apply in this case, although it is interesting. Here scope is minimal. Includes 30 lines that fit on the screen. This name does not matter. An equally good name is `recursion_handler`. Creating this method has no practical purpose.The method does not form part of the public interface of any class, module. I think that we spend too much time on small matters. If you have a different opinion, I invite you to send a separate PR, which will change this name.

Reply:
> This article does not apply in this case, although it is interesting. 
In my opinion it absolutely applies to this case. More on that below.

> Here scope is minimal - 30 lines that fit on the screen (including 7 lines of documentation). This name does not matter. 
The scope in the article's example is even smaller.

> An equally good name is recursion_handler. 
No, `recursion_handler` would never be a good name for a function/method, because it's a noun. A method name should be a verb.

> Creating this method has no practical purpose.The method does not form part of the public interface of any class, module. Names are important when part of the public interface only. 
I disagree. Names are always important, because as a developer in this project I don't only read code which is part of the public interface. I want to understand all of the code and names are very important in understanding.
Moreover, the method in the article in also not part of the public interface, because it has the `protected` modifier keyword.
 
> I think that we spend too much time on small matters. 
I don't think naming is a small matter. After all, it's one of the two hardest things in computer science.

On the one hand you say you agree with the article (I assume, as you say it was "interesting"), but on the other hand you don't want to apply it here, even though to me both situations are extremely similar.

Reply:
Protected methods are part of a publicly accessible interface if the class is not final. The class can be extended and the protected method can be called by external developer. In this case, there is no technical possibility of calling this method from a place other than the one you see on the screen.